src/java/org/apache/cassandra/config/GuardrailsOptions.java

```diff
@@ -0,0 +1,82 @@
+
+import org.apache.cassandra.db.guardrails.Guardrails;
+import org.apache.cassandra.db.guardrails.GuardrailsConfig;
+
+/**
+ * Configuration settings for guardrails populated from the Yaml file.
+ *
+ * <p>Note that the settings here must only be used to builf the {@link GuardrailsConfig} class and not directly by the
+ * code checking each guarded constraint. That code should use the higher level abstractions defined in
+ * {@link Guardrails}).
+ *
+ * <p>This contains a main setting, {@code enabled}, controlling if guardrails are globally active or not, and
+ * individual settings to control each guardrail.
+ *
+ * <p>We have 2 variants of guardrails, soft (warn) and hard (fail) limits, each guardrail having either one of the
+ * variants or both. Note in particular that hard limits only make sense for guardrails triggering during query
+ * execution. For other guardrails, say one triggering during compaction, failing does not make sense.
+ *
+ * <p>Additionally, each individual setting should have a specific value (typically -1 for numeric settings),
+ * that allows to disable the corresponding guardrail.
+ *
+ * <p>For consistency, guardrails based on a simple numeric threshold should use the naming scheme
+ * {@code <what_is_guarded>_warn_threshold} for soft limits and {@code <what_is_guarded>_fail_threshold} for hard
+ * ones, and if the value has a unit, that unit should be added at the end (for instance,
+ * {@code <what_is_guarded>_fail_threshold_in_kb}). For "boolean" guardrails that disable a feature, use
+ * {@code <what_is_guarded_enabled}. Other type of guardrails can use appropriate suffixes but should start with
+ * {@code <what_is_guarded>}.
+ */
+public class GuardrailsOptions
+{
+    public Boolean enabled;
+    public Integer tables_warn_threshold;
+    public Integer tables_fail_threshold;
+    public Integer columns_per_table_fail_threshold;
+    public Integer secondary_indexes_per_table_fail_threshold;
+    public Integer materialized_views_per_table_fail_threshold;
+    public Set<String> table_properties_disallowed;
+    public Set<String> table_properties_ignored;
+    public Boolean user_timestamps_enabled;
+
+    public void applyConfig()
+    {
+        GuardrailsConfig config = GuardrailsConfig.instance;
+        set(enabled, config::setEnabled);
+        set(tables_warn_threshold, config::setTablesWarnThreshold);
+        set(tables_fail_threshold, config::setTablesFailThreshold);
+        set(columns_per_table_fail_threshold, config::setColumnsPerTableFailThreshold);
+        set(secondary_indexes_per_table_fail_threshold, config::setSecondaryIndexesPerTableFailThreshold);
+        set(materialized_views_per_table_fail_threshold, config::setMaterializedViewsPerTableFailThreshold);
+        set(table_properties_disallowed, config::setTablePropertiesDisallowed);
+        set(table_properties_ignored, config::setTablePropertiesIgnored);
+        set(user_timestamps_enabled, config::setUserTimestampsEnabled);
+    }
```

Comment:
> We are going to have three ways of defining the config: yaml, JMX and virtual tables. The idea is centralizing the state and validation of guardrails in GuardrailsConfig, whereas GuardrailsOptions is just a definition for parsing cassandra.yaml.

Why can't `GuardrailsOptions` be the one thing blocking it?  We do this in track warnings and support yaml and JMX; I don't see the argument why we need 2 different classes to handle validation.

> I know this pattern is not very usual, and more commonly we have duplicated validation in the setters and in applyConfig

This problem is one of things being "fixed" in https://issues.apache.org/jira/browse/CASSANDRA-15234.  By moving away from primitives to domain specific types, parsing now handles validation, and can make sure we don't duplicate the logic.

Also, this solution doesn't "solve" this problem, as it just moves the setters to a different class.

> wdyt?

Still feel the same way, I don't see how the second class solves the concerns you proposed as it just moves the concerns to the second class rather than being in the first.

> we would need to make all the attributes public and with snake case

One thing I don't get is why we don't just use Jackson with SnakeYAML handling the yaml parsing... with Jackson we can make the fields private and rely on setters... solving this whole problem...  Since SnakeYAML still does the parsing, we shouldn't have a regression.  I have worked on the yaml parsing logic we use for CASSANDRA-15234, this is rather low-hanging fruit.

Reply:
What I'm trying to do, in the absence of any improvements introduced by CASSANDRA-15234, is trying to keep the public fields in the class used to parse yaml used exclusively for that, parsing yaml. That way we don't have any other class besides `Config` constantly having to choose between using the public attributes and their setters. That ways any class around trying to change the config properties for guardrails has to do so through the setters on `GuardrailsConfig`, where there aren't any exposed attributes. The attributes in `GuardrailsOptions` are exposed, sure, but once we have built our `GuardrailsConfig` changing those doesn't have any effect.

Those setters in `GuardrailsConfig` are trying to avoid the problem of duplicated config validation that we usually have. Very commonly we have a validation done in `DatabaseDescriptor` when we read the yaml, and another validation in the JMX-exposed setters in classes like `StorageService` and such. Those two validations are commonly repeating the same checks, and there is a risk that we miss the proper checks in one of the places. 

For example, let's see the checks for validating `conf.track_warnings.coordinator_read_size`. We have one set of validation checks in `TrackWarnings.LongByteThreshold#validate` that is used when parsing yaml. This method ensures that the property is positive and that it's not greater that the abort threshold. Similar checks are done in [`StorageService#setCoordinatorLargeReadWarnThresholdKB`](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/service/StorageService.java#L6221-L6227), but this time the check for verifying that the warn threshold isn't greater than the abort threshold. As a result, a user can set a warn threshold that is greater than the abort threshold through JMX, but not through the yaml config. I'm trying to avoid these problems caused by duplicated validation by centralizing it in a single place, which is the setters without exposed attributes.

Of course I'd be happy to combine both `GuadrailsOptions` and `GuardrailsConfig` into a single class if we were able to keep the attributes private, with the proper casing and with annotations defining the name in the yaml, as we do with Jackson. But I think we don't have that option available, is this right? Is this something that is going to be done in CASSANDRA-15234?

Reply:
I am going to ignore this conversation until the "pluggable" one is over.  If we do have a pluggable API for config, then having a different class makes a ton of sense

Reply:
For me the current approach is confusing as it duplicates the configuration fields. It also create an issue for CASSANDRA-15254 where I am changing slightly the logic of `SettingsTable` to automatically discover nested configurations fields to minimize the amount of work required when new options are added. So far people forgot about the `Settings` VT and did not exposed the new configuration settings. I am not stuck on which approach we should use but it should be consistent within the configuration or it will simply make our life more painful. I am fully open to discuss it on the dev list.

Reply:
@blerer so that change in VTs would use the setters in the nested configuration fields, or would it directly write the attributes? I understand it should use the setters so we can place some field validation there, even if we are currently forced to make the attributes public because of SnakeYAML parsing.

Reply:
The problem is even for getter. As I would like to automatically had new settings which is not possible if the settings start to be placed in different classes. Setters is even more complex has the logic is spread in different places and we have to create some mappings manually.

Reply:
I have changed the patch to use the public fields in `GuardrailsOptions` as the base of everything, and moved also the setters with validation to `GuardrailsOptions`. This `GuardrailsOptions` implements the `GuardrailsConfig` interface. So, `GuardrailsOptions` is the default yaml-based implementation of `GuardrailsOptions`. 

The pluggability part is provided by `GuardrailsConfigProvider`, which can serve different implementations of `GuardrailsConfig` based on the `ClientState`. The interfaces for the guardrails themselves are no more.

There are also a pair of `GuardrailsConfigService`/`GuardrailsConfigServiceMBean` classes for taking care of the JMX connectivity. These offer mostly the same getters/setters than `GuardrailsOptions` but with JMX-friendly method signatures. That is, representing collections as CSV, etc.